tests/TestRunner/upgrade_test.py

```diff
@@ -259,10 +259,14 @@ def copy_clientlib_from_local_repo(self, version):
         dest_lib_file = self.download_dir.joinpath(version, "libfdb_c.so")
         if dest_lib_file.exists():
             return
+        # Avoid race conditions in case of parallel test execution by first copying to a temporary file
+        # and then renaming it atomically
+        dest_file_tmp = Path("{}.{}".format(str(dest_lib_file), random_secret_string(8)))
         src_lib_file = self.local_binary_repo.joinpath(version, "lib", "libfdb_c-{}.so".format(version))
         assert src_lib_file.exists(), "Missing file {} in the local old binaries repository".format(src_lib_file)
         self.download_dir.joinpath(version).mkdir(parents=True, exist_ok=True)
-        shutil.copyfile(src_lib_file, dest_lib_file)
+        shutil.copyfile(src_lib_file, dest_file_tmp)
```

Comment:
Did you consider using a file lock here?

Reply:
Using file locking would be an alternative approach. The problem is that this part of the script is executed only in the CI builds, so I cannot test it manually. So I just did it in the simplest way.